juju/juju
# [JUJU-1653] Add support for removing secrets

Add support for removing secrets. A new `secret-remove` hook command is added.
We also add a new "scope" attribute to the secret metadata - when the scoped entity is removed, so too are any associated secrets. Right now, the scope is not able to be set - it is hard wired to the application which owns the secret.

Secret deletion is done using bulk collection removes, to clean up all the dependent records like permissions etc. To allow for this, extra checks are added to prevent a secret being created when the owner or scoped entity is not alive. And the top level parent metadata record is deleted first, plus the secret access methods have an extra check added and will return not found if the secret is gone.

Attempts to access a non existent secret result in permission denied.

## Checklist


## QA steps

```
juju deploy juju-qa-dummy-source
juju exec --unit dummy-source/0 "secret-add data=foo"
secret:cbu6ssdr22hi5hqcv3pg
juju exec --unit dummy-source/0 "secret-get secret:cbu6ssdr22hi5hqcv3pg"
data: foo
juju exec --unit dummy-source/0 "secret-remove secret:cbu6ssdr22hi5hqcv3pg"
juju exec --unit dummy-source/0 "secret-get secret:cbu6ssdr22hi5hqcv3pg"
ERROR permission denied
```
check app deletion also deletes secret
```
juju exec --unit dummy-source/0 "secret-add data=foo"
secret:cbu702tr22hi5hqcv3q0
juju secrets
URI                          Revision  Rotate  Backend  Age
secret:cbu702tr22hi5hqcv3q0         1  never   juju     3 seconds ago  

juju remove-application dummy-source
removing application dummy-source
juju secrets
URI  Revision  Rotate  Backend  Age

```
